File: python/year_2015/day_06.py
"""
https://adventofcode.com/2015/day/6
"""


from pathlib import Path
import numpy as np
import re

def process(text, instruction, grid):
    r = re.compile(r"(\D+) (\d+),(\d+) through (\d+),(\d+)")

    for line in text.splitlines():
        instruction(*r.match(line).groups())

    return int(grid.sum())


def part_1(text, size=1000):
    def instruction(action, *coordinates):
        x1, y1, x2, y2 = map(int, coordinates)
        x2, y2 = (x2 + 1, y2 + 1)

        if action == "toggle":
            grid[y1:y2, x1:x2] = 1 - grid[y1:y2, x1:x2]
        elif action:
            grid[y1:y2, x1:x2] = action == "turn on"

    grid = np.zeros((size, size), bool)

    return process(text, instruction, grid)


def part_2(text, size=1000):
    def instruction(action, *coordinates):
        x1, y1, x2, y2 = map(int, coordinates)
        x2, y2 = (x2 + 1, y2 + 1)

        if action == "toggle":
            grid[y1:y2, x1:x2] += 2
        elif action == "turn on":
            grid[y1:y2, x1:x2] += 1
        else:
            grid[y1:y2, x1:x2] -= 1
            np.clip(grid[y1:y2, x1:x2], 0, None, grid[y1:y2, x1:x2])

    grid = np.zeros((size, size), int)

    return process(text, instruction, grid)


def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_06.txt").read_text()


def test_part_1_input():
    assert part_1(_read_input()) == 543903

def test_part_1_example_1():
    assert part_1('turn on 0,0 through 999,999') == 1000000

def test_part_1_example_2():
    assert part_1('toggle 0,0 through 999,0') == 1000

def test_part_1_example_3():
    assert part_1('turn off 499,499 through 500,500') == 0

def test_part_2_input():
    assert part_2(_read_input(), 1000) == 14687245

def test_part_2_example_1():
    assert part_2('turn on 0,0 through 0,0') == 1

def test_part_2_example_2():
    assert part_2('toggle 0,0 through 999,999') == 2000000

File: rust/year_2015/src/day_06.rs
pub fn part_1(input: &str) -> usize {
    let mut lights = vec![[false; 1000]; 1000];
    for (action, (i0, i1), (j0, j1)) in parse_input(input) {
        match action {
            TurnOn | TurnOff => lights[i0..=i1]
                .iter_mut()
                .for_each(|row| row[j0..=j1].fill(action == TurnOn)),
            Toggle => lights[i0..=i1]
                .iter_mut()
                .flat_map(|row| &mut row[j0..=j1])
                .for_each(|light| *light = !*light),
        }
    }
    lights.into_iter().flatten().filter(|&light| light).count()
}

pub fn part_2(input: &str) -> usize {
    let mut lights = vec![[0u8; 1000]; 1000];
    for (action, (i0, i1), (j0, j1)) in parse_input(input) {
        let action = action as i8;
        lights[i0..=i1]
            .iter_mut()
            .flat_map(|row| &mut row[j0..=j1])
            .for_each(|light| *light = light.saturating_add_signed(action));
    }
    lights
        .into_iter()
        .flatten()
        .map(|light| light as usize)
        .sum()
}

#[derive(PartialEq, Eq)]
enum Action {
    TurnOn = 1,
    TurnOff = -1,
    Toggle = 2,
}

use Action::*;

fn parse_input(input: &str) -> impl Iterator<Item = (Action, (usize, usize), (usize, usize))> {
    input.lines().map(|line| {
        let (action, ranges) = if let Some(ranges) = line.strip_prefix("turn on ") {
            (TurnOn, ranges)
        } else if let Some(ranges) = line.strip_prefix("turn off ") {
            (TurnOff, ranges)
        } else if let Some(ranges) = line.strip_prefix("toggle ") {
            (Toggle, ranges)
        } else {
            panic!("invalid action: {line}");
        };
        let (start, end) = ranges.split_once(" through ").unwrap();
        let (i0, j0) = start.split_once(',').unwrap();
        let (i0, j0) = (i0.parse().unwrap(), j0.parse().unwrap());
        let (i1, j1) = end.split_once(',').unwrap();
        let (i1, j1) = (i1.parse().unwrap(), j1.parse().unwrap());
        (action, (i0, i1), (j0, j1))
    })
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const INPUT: &str = include_str!("../test_data/day_06.txt");

    #[test_case(INPUT => 543903)]
    fn part_1(input: &str) -> usize {
        super::part_1(input)
    }

    #[test_case(INPUT => 14687245)]
    fn part_2(input: &str) -> usize {
        super::part_2(input)
    }
}
